master/internal/api/params.go

```diff
@@ -0,0 +1,35 @@
+package api
+
+import "math"
+
+// EffectiveOffset returns effective offset.
+func EffectiveOffset(reqOffset int, total int) (offset int) {
+	switch {
+	case reqOffset < -total:
+		return 0
+	case reqOffset < 0:
+		return int(math.Max(float64(total+reqOffset), 0))
+	default:
+		return reqOffset
+	}
+}
+
+// EffectiveLimit returns effective limit.
+// Input: non-negative offset and limit.
+func EffectiveLimit(limit int, offset int, total int) int {
```

Comment:
panic if offset is negative?

It appears that limit is allowed to be negative right?  It's not clear if that's true from the comment.

Reply:
I addressed both.